File: framework/languages/dotnet/e2e/RegistrationCodec/Client/Scenarios/RcB4CodecCoexistenceScenario.cs
// Verifies RC-B4 Codec Coexistence behavior.
using RegistrationCodec.Client.Support;
using RegistrationCodec.Shared;
using Zlink.HttpClient;

namespace RegistrationCodec.Client.Scenarios;

// RC-B4 verifies JSON, Protobuf, and MessagePack coexistence on one channel.
internal static class RcB4CodecCoexistenceScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient requester,
        ZLinkHttpClient evidenceServer)
    {
        await requester.Post("/codec/roundtrip").Async<CodecScenarioRes>();
        var evidence = (await evidenceServer.Post("/evidence/wait")
            .Body(new EvidenceWaitReq([
                "codec-request|codec=json", "codec-request|codec=protobuf", "codec-request|codec=msgpack"
            ]))
            .Async<string[]>()).Body;
        ZlinkStreamAssert.Ensure(
            EvidenceText.HasCodec(evidence, "json", "application/json")
            && EvidenceText.HasCodec(evidence, "protobuf", "application/x-protobuf")
            && EvidenceText.HasCodec(evidence, "msgpack", "application/x-msgpack"),
            "RC-B4 expected all codec evidence.");

        Console.WriteLine("scenario RC-B4 passed");
    }
}

File: framework/languages/node/e2e/RegistrationCodec/Client/Scenarios/RcB4CodecCoexistenceScenario.ts
// RC-B4: 한 root에서 여러 codec을 함께 사용한다 시나리오를 검증한다.
import type { CodecScenarioRes } from '../../Shared/messages';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

// RC-B4 verifies that JSON, Protobuf, and MessagePack coexist in one host-level codec registry.
export async function runRcB4(serverUrl: string): Promise<void> {
  const result = await postJson<CodecScenarioRes>(serverUrl, '/codec/roundtrip');
  ensure(result.json?.value === 'echo:rc-b1', 'RC-B4 JSON reply mismatch.');
  ensure(result.protobufValue?.includes('echo:rc-b2') === true, 'RC-B4 Protobuf reply mismatch.');
  ensure(result.protobufValue?.includes('content:application/x-protobuf') === true, 'RC-B4 Protobuf content type mismatch.');
  ensure(result.messagePackValue?.includes('echo:rc-b3') === true, 'RC-B4 MessagePack reply mismatch.');
  ensure(result.messagePackValue?.includes('content:application/x-msgpack') === true, 'RC-B4 MessagePack content type mismatch.');
  const evidence = await postJson<readonly string[]>(serverUrl, '/evidence/wait', {
    containsAll: ['codec-request|codec=json', 'codec-request|codec=protobuf', 'codec-request|codec=msgpack'],
    timeoutMilliseconds: 10_000
  });
  ensure(
    evidence.some((line) => line.includes('codec-request|codec=json') && line.includes('content=application/json')),
    'RC-B4 JSON evidence missing.'
  );
  ensure(
    evidence.some((line) => line.includes('codec-request|codec=protobuf') && line.includes('content=application/x-protobuf')),
    'RC-B4 Protobuf evidence missing.'
  );
  ensure(
    evidence.some((line) => line.includes('codec-request|codec=msgpack') && line.includes('content=application/x-msgpack')),
    'RC-B4 MessagePack evidence missing.'
  );
  console.log('scenario RC-B4 passed');
}
